akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -454,6 +465,29 @@ object ByteString {
       else new ByteStrings(bytestrings.take(last) :+ bytestrings(last).take1(restToTake), n)
     }
 
+    override def takeRight(n: Int): ByteString =
+      if (n <= 0) ByteString.empty
+      else if (n >= length) this
+      else takeRight0(n)
+
+    private def takeRight0(n: Int): ByteString = {
+      val byteStringsSize = bytestrings.size
+      @tailrec def findSplit(fullTakes: Int, remainingToTake: Int): (Int, Int) = {
+        val bs = bytestrings(byteStringsSize - fullTakes - 1)
+        if (bs.length > remainingToTake) (fullTakes, remainingToTake)
+        else findSplit(fullTakes + 1, remainingToTake - bs.length)
+      }
+
+      val (fullTakes, remainingToTake) = findSplit(0, n)
+
+      if (fullTakes == 0)
```

Comment:
Yes, you're right! Thanks for your feedback :)
So we can rewrite the logic like this?
```
private def takeRight0(n: Int): ByteString = {
  val byteStringsSize = bytestrings.size
  @tailrec def takeRightWithFullTakesAndRemaining (fullTakes: Int, remainingToTake: Int): ByteString = {
    val bs = bytestrings(byteStringsSize - fullTakes - 1)
    if (bs.length > remainingToTake) {
      if (fullTakes == 0)
        bytestrings(byteStringsSize - 1).takeRight(n)
      else if (remainingToTake == 0)
        new ByteStrings(bytestrings.takeRight(fullTakes), n)
      else
        new ByteStrings(bytestrings(byteStringsSize - fullTakes - 1).takeRight1(remainingToTake) +: bytestrings.takeRight(fullTakes), n)
    }
    else findSplit(fullTakes + 1, remainingToTake - bs.length)
  }

  takeRightWithFullTakesAndRemaining(0, n)
}
```